### DISCUSSION TITLE
Partial screen rendering - more than single line but less then entire screen?

### DISCUSSION THREAD

**Discussion by yanshay** [slint-ui/slint#4251] [Q&A]
Currently for MCU it's possible to do either single line rendering or full screen rendering.
I'm using ESP32 and full screen doesn't fit in memory so I'm using single line rendering, but I could find memory for quarter of the display. 

I assume that if I could render in 4 chunks instead of 320 lines, rendering in total would take less time - am I correct with this assumption?

Is there a way to implement rendering for multiple lines, then send them to the display, then render next area and send and so on?
With DMA it could even be done in parallel with double buffering.

**Comment by ogoffart** [comment]
Currently we only support partial rendering one line at the time.
Not sure if the difference of rendering speed would be much better for bigger buffer.

That said, if you want to batch uploads of bigger buffer, you can render several lines in your buffer before uploading it.

> Is there a way to implement rendering for multiple lines, then send them to the display, then render next area and send and so on?
> With DMA it could even be done in parallel with double buffering.

Yes, this is what we are doing in the rp2040:
We have two lines buffer in memory, and we do the following:
 1. render the line into the rendering buffer
 2. wait for the previous DMA operation to be finished.
 3. swap the rendering buffer and the other buffer
 4. setup a DMA operation to send that buffer (where we just rendered the line into at step1) to the screen asynchronously
 
So step 4 happens in parallel with the next step1

See code around https://github.com/slint-ui/slint/blob/master/examples/mcu-board-support/pico_st7789.rs#L402

**Reply by yanshay** [reply]
> That said, if you want to batch uploads of bigger buffer, you can render several lines in your buffer before uploading it.

Every rendered line could be of different width, right? at least this is how it looks like from the API but maybe underneath I can assume all line widths in a single redraw would be the same?

Is it possible to make the one line render always generate full screen width?

**Reply by ogoffart** [reply]
In the current implementation, all lines will be the same width. Because we only compute rectangular dirty region. But we plan in the future to have more smart regions so this may change .

> Is it possible to make the one line render always generate full screen width?

If you disable the partial rendering by using the RepaintBufferType::NewBuffer, but this is a bit of a waste

What is your screen interface that needs full width lines?

**Reply by yanshay** [reply]
> > Is it possible to make the one line render always generate full screen width?
> 
> If you disable the partial rendering by using the RepaintBufferType::NewBuffer, but this is a bit of a waste

Interesting, do you recognize that rows don't change across different lines that come one after the other in the one line rendering option? That's nice.  

> What is your screen interface that needs full width lines?

My interface supports different widths. It's just that I'm using ESP32S3 no_std, and want to use DMA (there isn't a driver yet in no_std, so I'll need to setup all registers and dma myself). One of the issues I think is that if line changes then I need to send commands on every line, and would need to route the pins on every line, which will reduce efficiency and introduce complexity, at least in my app which does mostly full (or close to fulll) screen width updates.
